File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> wordSet(wordList.begin(),wordList.end());
        wordSet.erase(beginWord);
        queue<pair<string,int>> q;
        q.push({beginWord,1});
        int wordLen = (int)beginWord.size();
        while (!q.empty()){
            auto [word,dist] = q.front();
            q.pop();
            if (word == endWord)
                return dist;
            for (int i=0; i<wordLen; i++){
                char ogChar = word[i];
                for (char ch='a'; ch<='z'; ch++){
                    word[i] = ch;
                    if (wordSet.contains(word)){
                        wordSet.erase(word);
                        q.push({word,dist+1});
                    }
                }
                word[i] = ogChar;
            }
        }
        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordSet = set(wordList)
        if beginWord in wordSet:
            wordSet.remove(beginWord)
        q = deque()
        q.append((beginWord,1))
        wordSize = len(beginWord)
        while len(q) != 0:
            (word,dist) = q.popleft()
            if word == endWord:
                return dist
            for i in range(wordSize):
                prefix = word[:i]
                suffix = word[i+1:]
                for i in range(26):
                    candidate = prefix + chr(ord('a')+i) + suffix
                    if candidate in wordSet:
                        wordSet.remove(candidate)
                        q.append((candidate,dist+1))
        return 0